internal/resources/gsi.go

```diff
@@ -0,0 +1,509 @@
+}
+
+// Create will create/drop/build/alter a primary or secondary index.
+func (g *GSI) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
+	var plan providerschema.GsiDefinition
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &plan)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	var ddl string
+
+	// create build index statement.
+	if !plan.BuildIndexes.IsNull() {
+		ddl = fmt.Sprintf(
+			"BUILD INDEX ON `%s`.`%s`.`%s`(%s)",
+			plan.BucketName.ValueString(),
+			plan.ScopeName.ValueString(),
+			plan.CollectionName.ValueString(),
+			listStringValues(plan.BuildIndexes),
+		)
+	} else {
+		// create primary index statement.
+		if plan.IsPrimary.ValueBool() {
+			var indexName string
+			if !plan.IndexName.IsNull() {
+				indexName = plan.IndexName.ValueString()
+			} else {
+				indexName = "#primary"
+			}
+			ddl = fmt.Sprintf(
+				"CREATE PRIMARY INDEX `%s` ON `%s`.`%s`.`%s`  WITH { \"defer_build\": %t,  \"num_replica\": %d }",
+				indexName,
+				plan.BucketName.ValueString(),
+				plan.ScopeName.ValueString(),
+				plan.CollectionName.ValueString(),
+				plan.With.DeferBuild.ValueBool(),
+				plan.With.NumReplica.ValueInt64(),
+			)
+		} else {
+			// create secondary index statement.
+			ddl = fmt.Sprintf(
+				"CREATE INDEX `%s` ON `%s`.`%s`.`%s`(%s) ",
+				plan.IndexName.ValueString(),
+				plan.BucketName.ValueString(),
+				plan.ScopeName.ValueString(),
+				plan.CollectionName.ValueString(),
+				listStringValues(plan.IndexKeys),
+			)
+
+			if !plan.PartitionBy.IsNull() {
+				ddl += fmt.Sprintf(" PARTITION BY HASH(%s) ", listStringValues(plan.PartitionBy))
+			}
+
+			if !plan.Where.IsNull() {
+				ddl += fmt.Sprintf(" WHERE %s ", plan.Where.ValueString())
+			}
+
+			if !plan.PartitionBy.IsNull() {
+				withClause := fmt.Sprintf(
+					" WITH { \"defer_build\": %t,  \"num_replica\": %d,  \"num_partition\": %d } ",
+					plan.With.DeferBuild.ValueBool(),
+					plan.With.NumReplica.ValueInt64(),
+					plan.With.NumPartition.ValueInt64(),
+				)
+
+				ddl += withClause
+			} else {
+				// should not set num_partition for non-partitioned index.
+				withClause := fmt.Sprintf(
+					" WITH { \"defer_build\": %t,  \"num_replica\": %d} ",
+					plan.With.DeferBuild.ValueBool(),
+					plan.With.NumReplica.ValueInt64(),
+				)
+
+				ddl += withClause
+			}
+
+		}
+	}
+
+	if err := g.executeGsiDdl(ctx, &plan, ddl); err != nil {
+		resp.Diagnostics.AddError(
+			"An error occurred while executing index DDL",
+			"Error executing index DDL: "+ddl+"\nError: "+err.Error(),
+		)
+		return
+	}
+
+	resp.Diagnostics.Append(resp.State.Set(ctx, plan)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+}
+
+func (g *GSI) Read(ctx context.Context, req resource.ReadRequest, resp *resource.ReadResponse) {
+	var state providerschema.GsiDefinition
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	if !state.BuildIndexes.IsNull() {
+		return
+	}
```

Comment:
Can you please add a GoDoc comment explaining why we are returning here?

Reply:
Addressed in 5c6efe534d7761f3957ee6e5d1301148aba9aa2c